### PR TITLE
Use 2x less memory for MoE (including the DisposableTensor small tool)

### PR DESCRIPTION
**Pull request by fzyzcjy**
UPDATE: latest PR is in https://github.com/sgl-project/sglang/pull/6147

Brainstorm: maybe we can use `the_tensor.set_(torch.empty((0,), device=...))` (e.g. if users do not need to access metadata of the tensor after this operation)

## 2025.04.21 Update

I observe something similar for forward_deepgemm_masked. Will update this PR containing that as well. For users who need it before this PR is extracted from my dev branch, please directly visit https://github.com/sgl-project/sglang/pull/5524

image

## Original text

~~Do NOT look at the code now - it is based on two-batch-overlap branch (b/c that branch merges the tools branches). I will re-do the things on master branch later.~~
Updated

Before: peak 1.8+1.8+1.8GB = 5.4GB

image

After: peak is 1.8+0.9GB = 2.7GB

image

04.12: Updated the code, and re-checked the memory profile which still looks reasonable

Pasted image 20250412091002

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 1]
```diff
@@ -266,16 +276,18 @@ def forward(self, hidden_states: torch.Tensor, router_logits: torch.Tensor):
...
+                self.w13_weight.shape[1],
+                device=hidden_states.device,
+                dtype=hidden_states.dtype,
+            )
         )
         gateup_output = self.grouped_gemm_runner(
             a=gateup_input,
             b=self.w13_weight,
-            c=gateup_output,
+            c=gateup_output_creator,
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/layer.py:290]: The data type in line 89 is incorrect. How about we set DisposibleTensor and TensorCreator as subclasses of torch.Tensor to avoid such tedious refactor?

Comment (fzyzcjy) [reply python/sglang/srt/layers/moe/ep_moe/layer.py:290]: looks good, will do that later

**Comment by fzyzcjy** [comment]
Title seems to be inaccurate - 2x memory save is due to multiple changes
